HeterogeneousCore/AlpakaInterface/src/alpaka/CachingAllocator.cc

```diff
@@ -0,0 +1,502 @@
+      bool debug)
+      : device_(device),
+        cachedBlocks_(config.maxBin + 1),
+        binGrowth_(config.binGrowth),
+        minBin_(config.minBin),
+        maxBin_(config.maxBin),
+        minBinBytes_(detail::power(binGrowth_, minBin_)),
+        maxBinBytes_(detail::power(binGrowth_, maxBin_)),
+        maxCachedBytes_(cacheSize(config.maxCachedBytes, config.maxCachedFraction)),
+        reuseSameQueueAllocations_(reuseSameQueueAllocations),
+        debug_(debug),
+        fillAllocations_(config.fillAllocations),
+        fillAllocationValue_(config.fillAllocationValue),
+        fillReallocations_(config.fillReallocations),
+        fillReallocationValue_(config.fillReallocationValue),
+        fillDeallocations_(config.fillDeallocations),
+        fillDeallocationValue_(config.fillDeallocationValue),
+        fillCaches_(config.fillCaches),
+        fillCacheValue_(config.fillCacheValue) {
+    if (debug_) {
+      std::osyncstream out(std::cerr);
+      out << "CachingAllocator settings\n"
+          << "  bin growth " << binGrowth_ << "\n"
+          << "  min bin    " << minBin_ << "\n"
+          << "  max bin    " << maxBin_ << "\n"
+          << "  resulting bins:\n";
+      for (auto bin = minBin_; bin <= maxBin_; ++bin) {
+        auto binSize = detail::power(binGrowth_, bin);
+        out << "    " << std::right << std::setw(12) << detail::as_bytes(binSize) << '\n';
+      }
+      out << "  maximum amount of cached memory: " << detail::as_bytes(maxCachedBytes_) << '\n';
+    }
+  }
+
+  template <typename TDev, typename TQueue>
+  CachingAllocator<TDev, TQueue>::~CachingAllocator() {
+    // This should never be called while some memory blocks are still live.
+    assert(totalLive_ == 0);
+    freeAllCached();
+  }
+
+  // Return a copy of the cache allocation status, for monitoring purposes.
+  template <typename TDev, typename TQueue>
+  typename CachingAllocator<TDev, TQueue>::CachedBytes CachingAllocator<TDev, TQueue>::cacheStatus() const {
+    return CachedBytes{totalFree_, totalLive_, totalRequested_};
+  }
+
+  // Fill a memory buffer with the specified byte value.
+  // If the underlying device is the host and the allocator is configured to support immediate
+  // (non queue-ordered) operations, fill the memory synchronously using std::memset.
+  // Otherwise, let the alpaka queue schedule the operation.
+  //
+  // This is not used for deallocation/caching, because the memory may still be in use until the
+  // corresponding event is reached.
+  template <typename TDev, typename TQueue>
+  void CachingAllocator<TDev, TQueue>::immediateOrAsyncMemset(Queue queue, Buffer buffer, uint8_t value) {
+    // Host-only
+    if (std::is_same_v<Device, alpaka::DevCpu> and not reuseSameQueueAllocations_) {
+      std::memset(buffer.data(), value, alpaka::getExtentProduct(buffer) * sizeof(alpaka::Elem<Buffer>));
+    } else {
+      alpaka::memset(queue, buffer, value);
```

Comment:
What if this `memset()` throws an exception from unrelated asynchronous error?

Reply:
This function is only used when allocating a buffer.
If the `memset` fails, the allocation fails, _etc._